src/sys/petsc_interface.cxx

```diff
@@ -0,0 +1,173 @@
+/*!***********************************************************************
+ * Classes to wrap PETSc matrices and vectors, providing a convenient
+ * interface to them. In particular, they will internally convert
+ * between BOUT++ indices and PETSc ones, making it far easier to set
+ * up a linear system.
+ *
+ **************************************************************************
+ * Copyright 2013 J. Buchanan, J.Omotani
+ *
+ * Contact: Ben Dudson, bd512@york.ac.uk
+ *
+ * This file is part of BOUT++.
+ *
+ * BOUT++ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+#include "mpi.h"
+#include <memory>
+#include <vector>
+
+#include <bout/mesh.hxx>
+#include <bout/paralleltransform.hxx>
+#include <bout/petsc_interface.hxx>
+#include <bout/petsclib.hxx>
+#include <bout/region.hxx>
+#include <bout_types.hxx>
+#include <boutcomm.hxx>
+
+#ifdef BOUT_HAS_PETSC
+
+// GlobalIndexer implementation
+
+bool GlobalIndexer::initialisedGlobal = false;
+IndexerPtr GlobalIndexer::globalInstance;
+Mesh* GlobalIndexer::globalmesh = nullptr;
+
+IndexerPtr GlobalIndexer::getInstance(Mesh* localmesh) {
+  // Check that the identity of bout::globals::mesh hasn't changed
+  // since last call. (Needed for unit tests)
+  if (bout::globals::mesh != globalmesh) {
+    globalmesh = bout::globals::mesh;
+    initialisedGlobal = false;
+  }
+  if (localmesh == globalmesh) {
+    if (!initialisedGlobal) {
+      globalInstance = std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+      initialisedGlobal = true;
+    }
+    return globalInstance;
+  } else {
+    return std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+  }
+}
+
+void GlobalIndexer::initialiseTest() {
+  registerFieldForTest(indices3D);
+  registerFieldForTest(indices2D);
+  registerFieldForTest(indicesPerp);
+}
+
+void GlobalIndexer::initialise() {
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+  // Communicate a second time to get any corner values
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+}
+
+Mesh* GlobalIndexer::getMesh() { return fieldmesh; }
+
+PetscInt GlobalIndexer::getGlobal(Ind2D ind) {
+  return static_cast<PetscInt>(indices2D[ind] + 0.5);
```

Comment:
You can use `std::round` instead of the trick with `0.5`. The `static_cast` is still needed to get rid of a warning about implicit conversions